File: my-solutions/0003-longest-substring-without-repeating-characters/solution.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if(s.length() == 0)
            return 0;
        if(s.length() == 1)
            return 1;
        HashMap<Character, Integer> H = new HashMap<Character, Integer>();
        int i = 0;
        int j = 1;
        int max = 1;
        H.put(s.charAt(i), i);
        while(j < s.length()){
            
            Integer temp = H.get(s.charAt(j));
            if(temp == null){
                if(j - i + 1 > max)
                    max = j - i + 1;
                H.put(s.charAt(j), j);
                j++;
            }
            else{
                int k = i;
                i = temp + 1;  
                while(k<i){
                    H.remove(s.charAt(k));
                    k++;
                }
            }
            
        }
        return max;
    }
}

File: my-solutions/0003-longest-substring-without-repeating-characters/solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if (size(s) == 0)
            return 0;
        unordered_map<char, int> char_map;
        int max_window_size = 1; // window size will be 1 at a minimum
        int j = 0;
        char_map[s[0]] = 1;
        for(int i = 1; i < size(s); i++) {
            // add s[i] to the map
            if(char_map.find(s[i]) == char_map.end()) {
                char_map[s[i]] = 1;
            } else {
                char_map[s[i]]++;
            }

            // if char_map[s[i]] > 1, keep removing s[j++] from char_map till char_map[s[i]] becomes == 1
            while(char_map[s[i]] > 1) {
                if(char_map[s[j]] == 1) {
                    char_map.erase(s[j]);
                } else if(char_map[s[j]] > 1) {
                    char_map[s[j]]--;
                }
                j++;
            }
            int curr_window_size = i - j + 1;
            // cout << "current window: " << i << ", " << j << endl;
            max_window_size = std::max(max_window_size, curr_window_size);

        }
        return max_window_size;
    }
};
